custom-targets/vertex-ai/model-deployer/render.go

```diff
@@ -0,0 +1,226 @@
+	endpointRegex = regexp.MustCompile("^projects/([^/]+)/locations/([^/]+)/endpoints/([^/]+)$")
+)
+
+// renderer implements the handler interface for performing a render.
+type renderer struct {
+	gcsClient         *storage.Client
+	aiPlatformService *aiplatform.Service
+	params            *params
+	req               *clouddeploy.RenderRequest
+}
+
+// process processes the Render params by generating the YAML representation of a
+// DeployModelRequest object.
+func (r *renderer) process(ctx context.Context) error {
+	fmt.Println("Processing render request")
+	res, err := r.render(ctx)
+	if err != nil {
+		fmt.Printf("Render failed: %v\n", err)
+		res := &clouddeploy.RenderResult{
+			ResultStatus:   clouddeploy.RenderFailed,
+			FailureMessage: err.Error(),
+		}
+		r.addCommonMetadata(res)
+		fmt.Println("Uploading failed render results")
+		rURI, err := r.req.UploadResult(ctx, r.gcsClient, res)
+		if err != nil {
+			return fmt.Errorf("error uploading failed render results: %v", err)
+		}
+		fmt.Printf("Uploaded failed render results to %s\n", rURI)
+		return err
+	}
+	r.addCommonMetadata(res)
+
+	fmt.Println("Uploading successful render results")
+	rURI, err := r.req.UploadResult(ctx, r.gcsClient, res)
+	if err != nil {
+		return fmt.Errorf("error uploading render results: %v", err)
+	}
+	fmt.Printf("Uploaded render results to %s\n", rURI)
+	return nil
+}
+
+func (r *renderer) render(ctx context.Context) (*clouddeploy.RenderResult, error) {
+	fmt.Printf("Downloading render input archive to %s and unarchiving to %s\n", srcArchivePath, srcPath)
+	inURI, err := r.req.DownloadAndUnarchiveInput(ctx, r.gcsClient, srcArchivePath, srcPath)
+	if err != nil {
+		return nil, fmt.Errorf("unable to download and unarchive render input: %v", err)
+	}
+	fmt.Printf("Downloaded render input archive from %s\n", inURI)
+
+	out, err := r.renderDeployModelRequest()
+	if err != nil {
+		return nil, fmt.Errorf("error rendering deploy model params: %v", err)
+	}
+
+	fmt.Printf("Uploading deployed model manifest.\n")
+
+	mURI, err := r.req.UploadArtifact(ctx, r.gcsClient, "manifest.yaml", &clouddeploy.GCSUploadContent{Data: out})
+	if err != nil {
+		return nil, fmt.Errorf("error uploading deployed model manifest: %v", err)
+	}
+
+	fmt.Printf("Uploaded deployed model manifest to %s\n", mURI)
+
+	return &clouddeploy.RenderResult{
+		ResultStatus: clouddeploy.RenderSucceeded,
+		ManifestFile: mURI,
+	}, nil
+}
+
+// renderDeployModelRequest generates a DeployModelRequest object and returns its definition as a yaml-formatted string
+func (r *renderer) renderDeployModelRequest() ([]byte, error) {
+
+	if err := applyDeployParams(r.params.configPath); err != nil {
+		return nil, fmt.Errorf("cannot apply deploy parameters to configuration file: %v", err)
+	}
+
+	// blank deployed model template
+	deployedModel := &aiplatform.GoogleCloudAiplatformV1DeployedModel{}
+
+	configuration, err := loadConfigurationFile(r.params.configPath)
```

Comment:
What happens when this returns nil, nil? Why is that allowed.

Reply:
The config file is optional so if its not found and `vertexAIConfigurationPath` is not specified nothing is returned and there is no error.